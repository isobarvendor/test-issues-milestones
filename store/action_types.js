export const LOGIN= `login`;
export const GET_TOKEN= `getToken`;
export const SIGNUP= `signup`;
export const GET_ACCOUNT= `getAccount`;

//ngps
export const GET_LIST_PRIZE= `getListPrize`;
export const GET_LIST_WALLET= `getListWallet`;
export const CHECK_ATTEMPT= `checkAttempt`;
export const SUBMIT_FORM= `submitForm`;
export const UPLOAD_FILE= `uploadFile`;
export const DELETE_FILE= `deleteFile`;
export const REDEEM_PRIZE=`redeemPrize`;
export const GET_MY_PRIZE=`getMyPrize`;
export const GET_PHONE = `getPhone`;
export const GET_USER_DATA = `getUserData`

//cms
export const FETCH_CMS_DATA= `fethCMSData`;
export const GET_LIST_WINNERS =`getListWinners`;
