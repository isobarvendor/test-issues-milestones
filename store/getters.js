import deepClone from 'deep-clone';

export default {
    getLoginAccount(state) {
        return deepClone(state.login);
    },
    getCMSContent(state) {
        return deepClone(state.CMSContent);
    },
    getCMSConfig(state) {
      return deepClone(state.config);
    },
    getListPrize(state) {
      return deepClone(state.listPrize);
    }
}
