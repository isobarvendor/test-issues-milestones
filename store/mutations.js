

//shared store are concatenated to the central store
//TO-DO: manage the shared store seperately by module
export default {

  SET_LOGIN_ACCOUNT:(state, data)=>{
        state.login = data;
    },

  SET_TOKEN:(state, data)=>{
      state.token = data;
  },
  SET_ERROR:(state, data)=>{
    state.errorLogin = data;
},
  SET_LIST_PRIZE:(state, data)=>{
    state.listPrize = data;
  },
  SET_LIST_WALLET:(state, data)=>{
    state.listWallet = data;
  },
  SET_FILE_AMAZON:(state, data)=>{
    state.fileUploaded = data;
  },
  SET_CONFIG:(state,data)=>{
    state.config = data
  },
  SET_CMS_CONTENT:(state,data)=>{
    state.CMSContent = data
  }



};
