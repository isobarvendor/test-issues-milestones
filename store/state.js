export default () => ({
    details: {},
    productName: "coke",
    login:null,
    token:null,
    fileUploaded:null,
    listPrize:null,
    errorLogin:null,
    config:null,
    CMSContent:null,
    language:"en",
    redeemPrize:null,
    listWinners:null,
    listWallet:null,
    isCampaignStarted:false,
    goToSignin:false,
    fromBanCity:false,

});
