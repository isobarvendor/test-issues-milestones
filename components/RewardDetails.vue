<template>
  <div class="reward-details" v-if="data">
      <div class="back-icon">
         <img  @click="back()" v-if="$mq == 'xl'" src="/img/icons/back.png" />
           <img v-if="$mq == 'sm' || $mq == 'md' || $mq == 'lg'" src="/img/icons/close.png"/>
      </div>
     <div class="reward-body" >
         <div class="title">
          <h3>{{data.name}}</h3>
         </div>
         <p v-html="data.description"></p>
         <div class="redeem-expiry">
             <h4>Expiry Date</h4>
             <p v-html="data.shortDescription"></p>
         </div>
         <div class="redeem-expiry">
             <h4>How To  Redeem</h4>
             <p v-html="data.redeemDescription"></p>
         </div>
         <div class="error-message" v-if="errorMessage" v-html="errorMessage"></div>
         <div class="success-message" v-if="successMessage" v-html="successMessage"></div>
         <a class="button rewards-bottom" @click="redeemPrize(data.prizeId)"> Redeem ({{data.amountAvailable}} coins)</a>

    </div>
    <div class="image-fluid">
        <img :src="data.imgUrl"/>
        <a v-if="$mq == 'sm' || $mq == 'md'" class="button center mobile" @click="redeemPrize(data.prizeId)"> Redeem ({{data.amountAvailable}}  coins)</a>
    </div>

  </div>
</template>

<script>
import {REDEEM_PRIZE} from '@/store/action_types';
export default {
  name: "RewardDetails",
  components: {
  },
   props: {
    data: null,
  },
  data() {
    return {
      errorMessage:null,
      successMessage:null
    };
  },

  computed:{
  },
  methods:{
    back(){
      this.$router.replace('/rewards');
    },
    redeemPrize(prizeId){
      let request = {"prizeId":prizeId};
      this.$store.dispatch(REDEEM_PRIZE,request).then((response)=>{
         this.errorMessage=null;
         this.successMessage="Congratulations you already redeemed this prize";
       }).catch(error=>{
         if(error.response && error.response.data.status=="401"){
           localStorage.clear();
         }else{
            this.errorMessage="Sorry you can not redeem this prize";
         }
       })
    }

  },
  beforeMount() {},
  mounted(){

  }
};
</script>

<style lang="scss">
.reward-details{
    display: flex;
    flex-direction: row;
    width: 100%;
    .error-message{
      color:red;
      margin-top: 30px;
    }
    .success-message{
      color:green;
      margin-top: 30px;
    }
    .back-icon{
        display: flex;
        align-self: flex-start;
        margin: 120px 0;
        width:25%;
        justify-content: flex-end;
    }
    .reward-body{
        display: flex;
        flex-direction: column;
        width: 70%;
        margin: 120px 0px;
        padding: 0px 15px;
        .title{
            font-size: 30px;
            margin-bottom: 40px;
        }
        .redeem-expiry{
            margin-top: 60px;
            border-bottom: 1px solid #585858;
            width: 70%;
            h4{
                font-size: 25px;
                margin-bottom: 30px;
            }
            p{
                padding-bottom: 30px
            }
        }
    }
    .image-fluid{
        img{
            height: 100vh;
        }
    }
    a.button.rewards-bottom{
        margin-top: 30px;
    }
    @media only screen and (max-width: 1199px) {
        flex-direction: column-reverse;
        width: 100%;
        position: relative;
        p{
            line-height: 24px;
        }
        .reward-body{
            width:100%;
            margin-top: 45px;
            margin-bottom: 0px;
            .title{
                font-size: 20px;
                margin-bottom: 20px;
            }
            .redeem-expiry{
                margin-top: 40px;
                border-bottom: 1px solid #585858;
                width: 100%;
                h4{
                    font-size: 20px;
                    margin-bottom: 20px;
                }
                p{
                    padding-bottom: 20px;
                    line-height: 24px;
                }
            }
        }
        .image-fluid{
            img{
                height: auto;
                width: 100%;
            }
        }
        a.button.rewards-bottom{
            margin:auto;
            margin-top:40px;
        }
        a.button.rewards-bottom{
            margin:auto;
            margin-top:40px;
        }
        a.button.mobile{
            margin:auto;
            margin-top:-80px;
            position: relative;
            z-index: 10;
        }
        .back-icon{
            position: absolute;
            top: 0px;
            right:0px;
            margin-top: 10px;
        }
    }
}
</style>
