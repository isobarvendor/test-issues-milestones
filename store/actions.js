
import GeneralAPI from '../api/general';

import {LOGIN, SIGNUP, GET_ACCOUNT} from './action_types';

export default {
  [LOGIN]: ({ commit, state, getters }, data) => {
    return new Promise((resolve, reject) => {
      const moduleState = state;
      if (moduleState.login) {
        return resolve(moduleState.login)
      } else {
        GeneralAPI.login(data)
        .then(response => {
          commit('SET_TOKEN', response.data.token);
          sessionStorage.token=response.data.token;
          return resolve(messages);
        })
        .catch(error => {

          return reject(error);
        });
      }
    });
  },
  [GET_ACCOUNT]: ({ commit, state, getters }, token) => {
    return new Promise((resolve, reject) => {
      const moduleState = state;
      if (moduleState.login) {
        return resolve(moduleState.login)
      } else {
        GeneralAPI.getAccount(token)
        .then(response => {

          commit('SET_LOGIN_ACCOUNT', response.data);
          return resolve(messages);
        })
        .catch(error => {

          return reject(error);
        });
      }
    });
  },
  [SIGNUP]: ({ commit, state, getters }, data) => {
    return new Promise((resolve, reject) => {
      const moduleState = state;
      if (moduleState.login) {
        return resolve(moduleState.login)
      } else {
        GeneralAPI.signup(data)
        .then(messages => {
          //commit('SET_LOGIN_ACCOUNT', messages);
          return resolve(messages);
        })
        .catch(error => {
          console.error(error);
          return reject(error);
        });
      }
    });
  },


  /*[SET_GLOBAL_ERROR_DIALOG]:({commit, state}, status)=>{
    commit('SET_GLOBAL_ERROR_DIALOG', status);
  }
  */
};
