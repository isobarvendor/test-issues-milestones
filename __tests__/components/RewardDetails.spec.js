
import { mount,createLocalVue } from '@vue/test-utils'

import RewardDetails from "@/components/RewardDetails";


import mutations from '@/store/mutations';



import Vuex from 'vuex';

import {REDEEM_PRIZE} from '@/store/action_types';


Object.defineProperty(window, 'location', {
  writable: true,
  value: { assign: jest.fn() }
});

let localVue, store,actions;
let $mq ="xl";

actions = {
  [REDEEM_PRIZE]: jest.fn()
}

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.use(Vuex);
    store = new Vuex.Store({
      state:{
        language: "id",
        listWallet:{
          walletStatus:[
            {amount :130}
          ]
        }
      },
      actions,
      mutations
    });
  });

  const mockRouter = {
    replace: jest.fn()
  }

const factory = () => {
  return mount(RewardDetails, {
    mocks: {
      $mq:$mq,
      $router: mockRouter
    },
    propsData: {
      data :{
       name:"Reward1",
       shortDescription:"shortdescription",
       redeeemDescription:"shortdescription",
       prizeId:"prizeid",
       prizeCost:[{amount :10}],
      }
    },
    localVue, store
  });
};

describe("RewardDetails", () => {

  test("test all computed value", () => {
    const wrapper = factory();
    expect(wrapper.vm.listWallet).toBe(store.state.listWallet);
  });

  test("test back button", async() => {

    const wrapper = factory();
    wrapper.find('.back-icon').find('img').trigger('click');
    await expect(mockRouter.replace).toHaveBeenCalledWith('/rewards');

  });

  test("test can redeem function", async() => {

    const wrapper = factory();
    wrapper.vm.canRedeem(wrapper.vm.data.prizeCost);
    expect(wrapper.text().includes('coins')).toBe(true);

  });

  test("test redeem function", async() => {
    window.location.assign= jest.fn() // Create a spy

    const wrapper = factory();
    wrapper.find('.button').trigger('click');
    expect(store.state.redeemPrize).toBe(null);
    expect(actions[REDEEM_PRIZE]).toHaveBeenCalled();

   // expect(window.location.assign).toHaveBeenCalledWith("/congratulations");


  });




  test("mounts properly", () => {
    const wrapper = factory();
    expect(wrapper.vm).toBeTruthy();
  });

  test("renders properly", () => {
    const wrapper = factory();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
