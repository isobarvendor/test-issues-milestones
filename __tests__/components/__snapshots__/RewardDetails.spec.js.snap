// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RewardDetails renders properly 1`] = `
<div class="reward-details">
  <div class="back-icon"><img src="/img/icons/back.png">
    <!---->
  </div>
  <div class="reward-body">
    <div class="title">
      <h3>Reward1</h3>
    </div>
    <p></p>
    <div class="redeem-expiry">
      <h4>Expiry Date</h4>
      <p>shortdescription</p>
    </div>
    <div class="redeem-expiry">
      <h4>How To Redeem</h4>
      <p></p>
    </div>
    <!----> <a class="button rewards-bottom"> Redeem (10 coins)</a>
  </div>
  <div class="image-fluid"><img>
    <!---->
  </div>
</div>
`;
